extensions/gitpod-remote/src/extension.ts

```diff
@@ -0,0 +1,233 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Gitpod. All rights reserved.
+ *--------------------------------------------------------------------------------------------*/
+
+/// <reference path='../../../src/vscode-dts/vscode.d.ts'/>
+
+import * as cp from 'child_process';
+import { GitpodExtensionContext, registerTasks, setupGitpodContext, registerIpcHookCli } from 'gitpod-shared';
+import * as path from 'path';
+import * as util from 'util';
+import * as vscode from 'vscode';
+
+let gitpodContext: GitpodExtensionContext | undefined;
+export async function activate(context: vscode.ExtensionContext) {
+	gitpodContext = await setupGitpodContext(context);
+	if (!gitpodContext) {
+		return;
+	}
+
+	logContextInfo(gitpodContext);
+
+	if (vscode.extensions.getExtension('gitpod.gitpod')) {
+		try {
+			await util.promisify(cp.exec)('code --uninstall-extension gitpod.gitpod');
+			vscode.commands.executeCommand('workbench.action.reloadWindow');
+		} catch (e) {
+			gitpodContext.logger.error('failed to uninstall gitpod.gitpod:', e);
+		}
+		return;
+	}
+	if (openWorkspaceLocation(gitpodContext)) {
+		return;
+	}
+
+	registerTasks(gitpodContext, (options) => vscode.window.createTerminal(options));
+	installInitialExtensions(gitpodContext);
+	registerHearbeat(gitpodContext);
+
+	registerCLI(gitpodContext);
+	// configure task terminals if Gitpod Code Server is running
+	if (process.env.GITPOD_THEIA_PORT) {
+		registerIpcHookCli(gitpodContext);
+	}
+
+	// For port tunneling we rely on Remote SSH capabilities
+	// and gitpod.gitpod to disable auto tunneling from the current local machine.
+	vscode.commands.executeCommand('gitpod.api.autoTunnel', gitpodContext.info.getGitpodHost(), gitpodContext.info.getInstanceId(), false);
+
+	// For collecting logs, will be called by gitpod-desktop extension;
+	context.subscriptions.push(vscode.commands.registerCommand('__gitpod.getGitpodRemoteLogsUri', () => {
+		return context.logUri;
+	}));
+
+	// TODO
+	// - auth?
+	// - .gitpod.yml validations
+	// - add to .gitpod.yml command
+	// - cli integration
+	//   - git credential helper
+	await gitpodContext.active;
+}
+
+export function deactivate() {
+	if (!gitpodContext) {
+		return;
+	}
+	return gitpodContext.dispose();
+}
+
+function logContextInfo(context: GitpodExtensionContext) {
```

Comment:
We should do it always in gitpod-shared. No need to duplicate it web and remote.